src/quacc/recipes/vasp/core.py

```diff
@@ -15,6 +22,104 @@
     from quacc.schemas._aliases.vasp import DoubleRelaxSchema, VaspSchema
     from quacc.utils.files import Filenames, SourceDirectory
 
+logger = logging.getLogger(__name__)
+
+
+@job
+def nscf_job(
+    atoms: Atoms,
+    prev_dir: SourceDirectory,
+    bandgap: float | None = None,
+    nbands_factor: float | None = None,
+    preset: str | None = "BulkSet",
+    kpoints_mode: Literal["uniform", "line"] = "uniform",
+    calculate_optics: bool = False,
+    **calc_kwargs,
+) -> VaspSchema:
+    """
+    Carry out a non-self-consistent field (NSCF) calculation.
+
+    Parameters
+    ----------
+    atoms
+        Atoms object.
+    prev_dir
+        Directory (str | path) of the static_job
+    bandgap
+        The band gap of the material.
+    nbands_factor
+        A factor used to adjust NBANDS provided vasprun.xml(.gz) exists in
+        prev_dir
+    preset
+        Preset to use from `quacc.calculators.vasp.presets`.
+    kpoints_mode
+        Type of k-points mode. Options are "uniform" or "line". "boltztrap" may
+        be implemented in future.
+    calculate_optics
+        Whether to calculate optical properties.
+    **calc_kwargs
+        Custom kwargs for the Vasp calculator. Set a value to
+        `None` to remove a pre-existing key entirely. For a list of available
+        keys, refer to [quacc.calculators.vasp.vasp.Vasp][].
+
+    Returns
+    -------
+    VaspSchema
+        Dictionary of results from [quacc.schemas.vasp.vasp_summarize_run][].
+        See the type-hint for the data structure.
+    """
+
+    calc_defaults = {
+        "lorbit": 11,
+        "lwave": False,
+        "lcharg": False,
+        "nsw": 0,
+        "isym": 0,
+        "icharg": 11,
+        "kspacing": None,
+        "nedos": 5001,
+    }
+
+    updates: dict[str, Any] = {}
+
+    # update NBANDS if vasprun.xml* in prev_dir exists
+    vasprun_exists = any(
+        file.startswith("vasprun.xml") for file in os.listdir(prev_dir)
+    )
+    if vasprun_exists:
+        vasprun_path = Path(prev_dir, "vasprun.xml")
+        vasprun = Vasprun(zpath(vasprun_path))
+        if nbands_factor is not None:
+            nbands = int(np.ceil(vasprun.parameters["NBANDS"] * nbands_factor))
+            updates["nbands"] = nbands
+    else:
+        logger.warning("vasprun.xml* file does not exist in the specified directory.")
+
+    if kpoints_mode == "uniform":
+        # Use tetrahedron method for DOS and optics calculations
+        updates.update({"ismear": -5, "isym": 2})
+    elif kpoints_mode == "line":
+        sigma = 0.2 if bandgap < 1e-4 else 0.01
+        updates.update({"ismear": 0, "sigma": sigma})
+    else:
+        raise ValueError("Supported kpoint modes are 'uniform' and 'line' at present")
+
+    if calculate_optics:
+        updates.update({"loptics": True, "lreal": False, "cshift": 1e-5})
+
+    # integrate updates to calc_kwargs
+    calc_kwargs.update(updates)
```

Comment:
These updates should be added to the `calc_defaults` instead. The reason for this is we want to always respect the user. If the user chose to do `NBANDS = 10000`, then we don't want to override their decision. In `base_fn()`, `calc_swaps` is merged with `calc_defaults`, with the former taking precedent.